### DISCUSSION TITLE
This is Awesome! ✨️

### DISCUSSION THREAD

**Discussion by zicklag** [risc0/risc0#51] [General]
Hey there, I just found this and I think it's an awesome idea. I know it's not ready for anything yet, and I'm not sure I'll be able to use it for anything, but it's a brilliant idea if it actually works.

Very cool. :smile:

**Comment by brianretford** [comment]
It does work! And it will be coming to various blockchains sometime ;-)

**Reply by zicklag** [reply]
I'm glad to hear that!

I'm entertaining the idea of creating a new blockchain using Substrate that's essentially a modified version of the https://terra.money stable coin algorithm, but I want to add privacy to the list of features. 

I'm not personally super deep in the math or low-level cryptography, and I've been browsing around trying to understand what options are available for zero-knowledge proofs.

I _really_ like the concept of risc-0, but I'm not yet sure if it will work for my use-case.

Is it possible to run risk-0 validations inside of WASM? I think that's the biggest limitation I have because I'm building on substrate, which runs the state-transition-function in WASM.

**Reply by brianretford** [reply]
Indeed it is! We're working on a number or related ideas :-)

**Reply by zicklag** [reply]
OK, then I have to seriously look into this! :detective:

I'll open new discussions if I have more questions. Thanks!

**Reply by justinfrevert** [reply]
It's possible - I started making some basic examples https://github.com/justinFrevert/substrate-risc0

**Reply by vedhavyas** [reply]
@justinfrevert This is very cool! Before I try and modify the factors code, I want to check if this is possible to prove the WASM execution as well

**Reply by justinfrevert** [reply]
@vedhavyas a lot more work needs to be done around the runtime apis and host functions before we can do that. I think Substrate wasn't really built with proving the WASM execution in-mind.

**Reply by vedhavyas** [reply]
Thank you for the quick response.

That is fair but this can definitely be useful when the they are verifying the parachain blocks on relay chain I believe

**Reply by vedhavyas** [reply]
I have also come across this example - https://github.com/risc0/risc0/tree/main/examples/wasm
If Substrate WASM can be initialized properly, shouldn't this approach not work. 
I could be missing something here as well

**Reply by flaub** [reply]
We have not yet gotten proving to work within WASM, this will require quite a bit more work before it's ready. However, verification should work fine within WASM.
